exonum/src/api/mod.rs

```diff
@@ -71,10 +82,62 @@ pub trait ApiBackend: Sized {
         E: Into<With<Q, I, R, F>>,
         Self::Handler: From<NamedWith<Q, I, R, F, Mutable>>,
     {
-        let named_with = NamedWith::new(name, endpoint);
+        let named_with = NamedWith::new(name, endpoint, Actuality::Actual);
+        self.raw_handler(Self::Handler::from(named_with))
+    }
+
+    /// Adds the given endpoint handler to the backend, marking it as deprecated.
+    fn deprecated_endpoint<N, Q, I, R, F, E>(
+        &mut self,
+        name: N,
+        discontinued_on: Option<Date<Utc>>,
+        endpoint: E,
+    ) -> &mut Self
+    where
+        N: Into<String>,
+        Q: DeserializeOwned + 'static,
+        I: Serialize + 'static,
+        F: Fn(Q) -> R + 'static + Clone,
+        E: Into<With<Q, I, R, F>>,
+        Self::Handler: From<NamedWith<Q, I, R, F, Immutable>>,
+    {
+        let named_with = NamedWith::new(name, endpoint, Actuality::Deprecated(discontinued_on));
+        self.raw_handler(Self::Handler::from(named_with))
+    }
+
+    /// Adds the given mutable endpoint handler to the backend, marking it as deprecated.
+    fn deprecated_endpoint_mut<N, Q, I, R, F, E>(
+        &mut self,
+        name: N,
+        discontinued_on: Option<Date<Utc>>,
+        endpoint: E,
+    ) -> &mut Self
+    where
+        N: Into<String>,
+        Q: DeserializeOwned + 'static,
+        I: Serialize + 'static,
+        F: Fn(Q) -> R + 'static + Clone,
+        E: Into<With<Q, I, R, F>>,
+        Self::Handler: From<NamedWith<Q, I, R, F, Mutable>>,
+    {
+        let named_with = NamedWith::new(name, endpoint, Actuality::Deprecated(discontinued_on));
         self.raw_handler(Self::Handler::from(named_with))
     }
 
+    /// Creates an endpoint which will return "301 Moved Permanently" HTTP status code
+    /// to the incoming requests.
+    /// Response will include a "Location" header denoting a new location of the resource.
+    fn moved_permanently(
+        &mut self,
+        name: &'static str,
+        new_location: &'static str,
```

Comment:
I think user agents such as browsers [redirect to the provided location](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/301); i.e., it should be specific for this particular request. Hence, I think that service developers should be provided with a handler instead of a static replacement. Static replacement could make more sense with the `Link` header with `rel="latest-version"` / `rel="alternate"` as per the [`Deprecation` header draft](https://tools.ietf.org/html/draft-dalal-deprecation-header-01).

Reply:
From [RFC 7231](https://tools.ietf.org/html/rfc7231#section-6.4.2):

>The server SHOULD generate a Location header field in the response
   containing a preferred URI reference for the new permanent URI.  The
   user agent MAY use the Location field value for automatic
   redirection.

Reply:
So, I guess, `MovedPermanently` is intended for basically just redirecting to a new location without any changes (even changing the request type is not recommended), and current signature suitable for such a convention. Or am I wrong here?